src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,597 @@
+    ) {
+      revert PPT_TotalCostBelowTolerance();
+    }
+
+    uint markValueToStrikeDiffRatio =
+      (lowerStrike.markPrice - higherStrike.markPrice).abs().divideDecimal(higherStrike.strike - lowerStrike.strike);
+
+    /*
+     * This check ensures that the (markPrice difference) / (strike Difference) ratio is within a certain range.
+     * Typically, the range will be between ~.4 and ~.6, but this can be adjusted in the params.
+     * If the ratio is too high then we are buying a spread that is too expensive in respect to the strikes.
+     * If too low, the trade is not as risky. This is to force the vault to only make realistic slightly-risky trades.
+     */
+    if (
+      markValueToStrikeDiffRatio < $.ppParams.minMarkValueToStrikeDiffRatio
+        || markValueToStrikeDiffRatio > $.ppParams.maxMarkValueToStrikeDiffRatio
+    ) {
+      revert PPT_MarkValueNotWithinBounds();
+    }
+  }
+
+  function _getOptionPrice(uint optionExpiry, uint optionStrike)
+    internal
+    view
+    returns (uint callPrice, uint putPrice, uint callDelta)
+  {
+    uint timeToExpiry = optionExpiry - block.timestamp;
+    (uint vol, uint forwardPrice) = _getFeedValues(optionStrike.toUint128(), optionExpiry.toUint64());
+
+    return Black76.pricesAndDelta(
+      Black76.Black76Inputs({
+        timeToExpirySec: timeToExpiry.toUint64(),
+        volatility: vol.toUint128(),
+        fwdPrice: forwardPrice.toUint128(),
+        strikePrice: optionStrike.toUint128(),
+        discount: 1e18
+      })
+    );
+  }
+
+  function _getFeedValues(uint128 strike, uint64 expiry) internal view returns (uint vol, uint forwardPrice) {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    StandardManager srm = StandardManager(address(getBaseTSAAddresses().manager));
+    IStandardManager.AssetDetail memory assetDetails = srm.assetDetails($.optionAsset);
+    (, IForwardFeed fwdFeed, IVolFeed volFeed) = srm.getMarketFeeds(assetDetails.marketId);
+    (vol,) = volFeed.getVol(strike, expiry);
+    (forwardPrice,) = fwdFeed.getForwardPrice(expiry);
+  }
+
+  ///////////////////
+  // Account Value //
+  ///////////////////
+
+  function _getSubAccountStats() internal view returns (uint openSpreads, uint baseBalance, int cashBalance) {
+    BaseTSAAddresses memory tsaAddresses = getBaseTSAAddresses();
+    ISubAccounts.AssetBalance[] memory balances = tsaAddresses.subAccounts.getAccountBalances(subAccount());
+    for (uint i = 0; i < balances.length; i++) {
+      if (balances[i].asset == _getPPTSAStorage().optionAsset && balances[i].balance > 0) {
+        // since we validate that we have a 1:1 ratio between buying and selling between two different strikes,
+        // we can just add the positive balances.
+        // We just validate that spreads are either open or how much they can lose,
+        // so we only care about one side of the balances.
+        openSpreads += balances[i].balance.toUint256();
```

Comment:
nit: safe cast isn't necessary as you check balance > 0 already

May help with contract size to just do `uint(balances[i].balance)`

Reply:
Sounds good wasn't sure what the norm is. Ill change this wherever I can.